Can't backup/restore workspaces because of old rsync version

Got this errors:
```
2017-02-21 00:33:34,539[ceSharedPool-12]  [ERROR] [DockerEnvironmentBackupManager 509]  - Error occurred during backup/restore of workspace 'workspace*************' on node 'node7.codenvy.io' : Warning: Permanently added '[node7.codenv
rsync: on remote machine: --usermap=*:1000: unknown option
rsync error: syntax or usage error (code 1) at main.c(1435) [server=3.0.9]
rsync: connection unexpectedly closed (0 bytes received so far) [sender]
rsync error: error in rsync protocol data stream (code 12) at io.c(226) [sender=3.1.2]
2017-02-21 00:33:34,540[ceSharedPool-12]  [ERROR] [DockerEnvironmentBackupManager 319]  - Synchronization process failed. Exit code 12
java.io.IOException: Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeCommand(DockerEnvironmentBackupManager.java:511) [codenvy-machine-hosted-5.3.0.jar:5.3.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.restoreBackupInsideLock(DockerEnvironmentBackupManager.java:308) [codenvy-machine-hosted-5.3.0.jar:5.3.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.restoreWorkspaceBackup(DockerEnvironmentBackupManager.java:194) [codenvy-machine-hosted-5.3.0.jar:5.3.0]
        at com.codenvy.machine.RemoteDockerNode.bindWorkspace(RemoteDockerNode.java:94) [codenvy-machine-hosted-5.3.0.jar:5.3.0]
        at com.codenvy.machine.agent.launcher.MachineInnerRsyncAgentLauncherImpl.launch(MachineInnerRsyncAgentLauncherImpl.java:68) [codenvy-machine-hosted-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.launchAgents(WorkspaceRuntimes.java:695) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$MachineAgentsLauncher.started(WorkspaceRuntimes.java:1048) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:816) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.start(CheEnvironmentEngine.java:267) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.startEnvironmentAndPublishEvents(WorkspaceRuntimes.java:714) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.access$100(WorkspaceRuntimes.java:106) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartTask.call(WorkspaceRuntimes.java:979) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StartTask.call(WorkspaceRuntimes.java:943) [che-core-api-workspace-5.3.0.jar:5.3.0]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalCallable.call(CopyThreadLocalCallable.java:30) [che-core-commons-lang-5.3.0.jar:5.3.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_111-internal]
```
This is because of old version of `rsync`, probably on old debian
Need to find way to work even on old debian or print message to user explaining why we can't start workspace.

Also please include more info in error message for second error, it probably related to previous one because of time difference is so small, but need to be sure.
So please include at least workspace ID in error message.

**Codenvy version:**    5.3.0

**OS and version:**    Centos 7.2

**Docker version:** 1.12.5

--- a/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
@@ -43,7 +43,7 @@ SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
 # If user in that container has access to sudo we use it.
 # Otherwise we use rsync without sudo and if it is possible ownership will be changed.
 RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${SRC_HOST} \
-            "if hash sudo 2>/dev/null;then echo 'sudo rsync';else echo 'rsync';fi")
+            "if hash sudo 2>/dev/null; then echo 'sudo rsync'; else echo 'rsync'; fi")
 
 ##### Rsync options #####
 RSYNC_OPTIONS=""

--- a/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncbackup.sh.erb
@@ -1,7 +1,5 @@
 #!/bin/bash
 
-SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
-
 SRC_FOLDER=${1}
 SRC_HOST=${2}
 SRC_PORT=${3}
@@ -9,14 +7,12 @@ DST_FOLDER=${4}
 REMOVE_ON_SUCCESS=${5=false}
 SRC_USER_NAME=${6}
 
-# ensure existing trailing slash to overwrite permissions of destination folder too
-if [[ ${SRC_FOLDER} != */ ]]; then
-   SRC_FOLDER=${SRC_FOLDER}"/"
-fi
-if [[ ${DST_FOLDER} != */ ]]; then
-   DST_FOLDER=${DST_FOLDER}"/"
-fi
-
+# Key for SSH connection to sync files
+SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
+# When Codenvy runs on windows we have issue since key doesn't have permissions 0600.
+# With another permissions on key SSH doesn't work.
+# So we copy key into another place that should be inside of container - so it is Linux and permissions are supported.
+# And apply needed permission to copied file. Copying is performed lazily, so it is skipped when it was done previously.
 # ensure ssh key has correct permissions
 if [[ ! -f /opt/rsync_key ]]; then
     cp $SSH_KEY /opt/rsync_key
@@ -25,16 +21,63 @@ if [[ $(stat -c %a /opt/rsync_key) != 600 ]]; then
     chmod 0600 /opt/rsync_key
 fi
 
+##### Set by puppet #####
 RSYNC_BACKUP_BWLIMIT=<%= scope.lookupvar('codenvy::rsync_backup_bwlimit') %>
 SSH_LOG_LEVEL=<%= scope.lookupvar('codenvy::rsync_ssh_log_level') %>
-SSH_OPTIONS="-i /opt/rsync_key -l ${SRC_USER_NAME} -o StrictHostKeyChecking=no -o PasswordAuthentication=no -o LogLevel=${SSH_LOG_LEVEL} -p ${SRC_PORT}"
+#########################
 
+##### SSH options #####
+SSH_OPTIONS=""
+# Add SSH connection options
+SSH_OPTIONS+=" -i /opt/rsync_key -l ${SRC_USER_NAME} -p ${SRC_PORT}"
+# Disable password authentication since we use key-based auth
+SSH_OPTIONS+=" -o PasswordAuthentication=no"
+# Disable hosts fingerprint checking because it may fail due to
+# starting different containers on the same ports after some period
+SSH_OPTIONS+=" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+# Set SSH logging level to make it possible to investigate problems
+SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
+#######################
+
+# We need root permissions in target container to change ownership of files.
+# If user in that container has access to sudo we use it.
+# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
 RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${SRC_HOST} \
             "if hash sudo 2>/dev/null;then echo 'sudo rsync';else echo 'rsync';fi")
 
-rsync --quiet --delete --partial --links --safe-links --recursive --times --human-readable --bwlimit=${RSYNC_BACKUP_BWLIMIT} --executability \
-              --include='.codenvy/' \
-              --filter=':- .gitignore' \
-              --rsh="ssh ${SSH_OPTIONS}" \
-              --rsync-path="${RSYNC_COMMAND}" \
-              ${SRC_HOST}:${SRC_FOLDER} ${DST_FOLDER}
+##### Rsync options #####
+RSYNC_OPTIONS=""
+RSYNC_OPTIONS+=" --quiet"
+RSYNC_OPTIONS+=" --recursive"
+# Throughput limit for rsync
+RSYNC_OPTIONS+=" --bwlimit=${RSYNC_BACKUP_BWLIMIT}"
+# Sync modification timestamps to optimise transfer of not modified files.
+RSYNC_OPTIONS+=" --times"
+# Do not remove partially transferred files if transfer is interrupted. Make subsequent transfers faster.
+RSYNC_OPTIONS+=" --partial"
+# Delete files/folders on receiving side if they are not present on sending side.
+RSYNC_OPTIONS+=" --delete"
+# Preserve sym links in a safe way
+RSYNC_OPTIONS+=" --links --safe-links"
+# Transition of ownership and permissions
+RSYNC_OPTIONS+=" --owner --group --numeric-ids --perms"
+#########################
+
+# Previously was used to overwrite permissions of root folders of syncing.
+# Was left for backward compatibility and to prevent possible issues after changing behavior.
+# Ensure existing trailing slash to overwrite permissions of destination folder too.
+if [[ ${SRC_FOLDER} != */ ]]; then
+   SRC_FOLDER=${SRC_FOLDER}"/"
+fi
+if [[ ${DST_FOLDER} != */ ]]; then
+   DST_FOLDER=${DST_FOLDER}"/"
+fi
+
+# Include .codenvy folders and respect .gitignore files.
+# Notice that not all the syntax of .gitignore is supported in rysnc (at least in v3.1.1).
+# TODO consider removing of these options, looks like they are not needed/reliable
+rsync ${RSYNC_OPTIONS} \
+      --rsh="ssh ${SSH_OPTIONS}" \
+      --rsync-path="${RSYNC_COMMAND}" \
+      --include='.codenvy/' --filter=':- .gitignore' \
+      ${SRC_HOST}:${SRC_FOLDER} ${DST_FOLDER}

--- a/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
@@ -43,7 +43,7 @@ SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
 # If user in that container has access to sudo we use it.
 # Otherwise we use rsync without sudo and if it is possible ownership will be changed.
 RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${DST_HOST} \
-            "if hash sudo 2>/dev/null;then echo 'sudo rsync';else echo 'rsync';fi")
+            "if hash sudo 2>/dev/null; then echo 'sudo rsync'; else echo 'rsync'; fi")
 
 ##### Rsync options #####
 RSYNC_OPTIONS=""
@@ -53,8 +53,6 @@ RSYNC_OPTIONS+=" --recursive"
 RSYNC_OPTIONS+=" --bwlimit=${RSYNC_RESTORE_BWLIMIT}"
 # Sync modification timestamps to optimise transfer of not modified files.
 RSYNC_OPTIONS+=" --times"
-# Do not remove partially transferred files if transfer is interrupted. Make subsequent transfers faster.
-RSYNC_OPTIONS+=" --partial"
 # Delete files/folders on receiving side if they are not present on sending side.
 RSYNC_OPTIONS+=" --delete"
 # Preserve sym links in a safe way

--- a/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
+++ b/dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb
@@ -1,7 +1,5 @@
 #!/bin/bash
 
-SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
-
 SRC_FOLDER=${1}
 DST_FOLDER=${2}
 DST_HOST=${3}
@@ -10,15 +8,12 @@ DST_USR_ID=${5}
 DST_GRP_ID=${6}
 SRC_USER_NAME=${7}
 
-# ensure existing trailing slash to overwrite permissions of destination folder too
-if [[ ${SRC_FOLDER} != */ ]]; then
-   SRC_FOLDER=${SRC_FOLDER}"/"
-fi
-if [[ ${DST_FOLDER} != */ ]]; then
-   DST_FOLDER=${DST_FOLDER}"/"
-fi
-
-# ensure ssh key has correct permissions
+# Key for SSH connection to sync files
+SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
+# When Codenvy runs on windows we have issue since key doesn't have permissions 0600.
+# With another permissions on key SSH doesn't work.
+# So we copy key into another place that should be inside of container - so it is Linux and permissions are supported.
+# And apply needed permission to copied file. Copying is performed lazily, so it is skipped when it was done previously.
 if [[ ! -f /opt/rsync_key ]]; then
     cp $SSH_KEY /opt/rsync_key
 fi
@@ -26,17 +21,63 @@ if [[ $(stat -c %a /opt/rsync_key) != 600 ]]; then
     chmod 0600 /opt/rsync_key
 fi
 
+##### Set by puppet #####
 RSYNC_RESTORE_BWLIMIT=<%= scope.lookupvar('codenvy::rsync_restore_bwlimit') %>
 SSH_LOG_LEVEL=<%= scope.lookupvar('codenvy::rsync_ssh_log_level') %>
-SSH_OPTIONS="-i /opt/rsync_key -l ${SRC_USER_NAME} -o StrictHostKeyChecking=no -o PasswordAuthentication=no -o LogLevel=${SSH_LOG_LEVEL} -p ${DST_PORT}"
+#########################
 
+##### SSH options #####
+SSH_OPTIONS=""
+# Add SSH connection options
+SSH_OPTIONS+=" -i /opt/rsync_key -l ${SRC_USER_NAME} -p ${DST_PORT}"
+# Disable password authentication since we use key-based auth
+SSH_OPTIONS+=" -o PasswordAuthentication=no"
+# Disable hosts fingerprint checking because it may fail due to
+# starting different containers on the same ports after some period
+SSH_OPTIONS+=" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+# Set SSH logging level to make it possible to investigate problems
+SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
+#######################
+
+# We need root permissions in target container to change ownership of files.
+# If user in that container has access to sudo we use it.
+# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
 RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${DST_HOST} \
             "if hash sudo 2>/dev/null;then echo 'sudo rsync';else echo 'rsync';fi")
 
-rsync --quiet --delete --partial --links --safe-links --owner --group --recursive --times --human-readable --bwlimit=${RSYNC_RESTORE_BWLIMIT} --executability \
-      --include='.codenvy/' \
-      --filter=':- .gitignore' \
+##### Rsync options #####
+RSYNC_OPTIONS=""
+RSYNC_OPTIONS+=" --quiet"
+RSYNC_OPTIONS+=" --recursive"
+# Throughput limit for rsync
+RSYNC_OPTIONS+=" --bwlimit=${RSYNC_RESTORE_BWLIMIT}"
+# Sync modification timestamps to optimise transfer of not modified files.
+RSYNC_OPTIONS+=" --times"
+# Do not remove partially transferred files if transfer is interrupted. Make subsequent transfers faster.
+RSYNC_OPTIONS+=" --partial"
+# Delete files/folders on receiving side if they are not present on sending side.
+RSYNC_OPTIONS+=" --delete"
+# Preserve sym links in a safe way
+RSYNC_OPTIONS+=" --links --safe-links"
+# Transition of ownership and permissions
+RSYNC_OPTIONS+=" --owner --group --numeric-ids --perms"
+#########################
+
+# Previously was used to overwrite permissions of root folders of syncing.
+# Was left for backward compatibility and to prevent possible issues after changing behavior.
+# Ensure existing trailing slash to overwrite permissions of destination folder too.
+if [[ ${SRC_FOLDER} != */ ]]; then
+   SRC_FOLDER=${SRC_FOLDER}"/"
+fi
+if [[ ${DST_FOLDER} != */ ]]; then
+   DST_FOLDER=${DST_FOLDER}"/"
+fi
+
+# Include .codenvy folders and respect .gitignore files.
+# Notice that not all the syntax of .gitignore is supported in rysnc (at least in v3.1.1).
+# TODO consider removing of these options, looks like they are not needed/reliable
+rsync ${RSYNC_OPTIONS} \
       --rsh="ssh ${SSH_OPTIONS}" \
       --rsync-path="${RSYNC_COMMAND}" \
-      --chown ${DST_USR_ID}:${DST_GRP_ID} \
+      --include='.codenvy/' --filter=':- .gitignore' \
       ${SRC_FOLDER} ${DST_HOST}:${DST_FOLDER}

